### ISSUE TITLE
Looping through data?

### ISSUE THREAD

**Issue by domfournier** [simpeg/simpeg#968]
https://github.com/simpeg/simpeg/blob/d20eefa6b2238210d1c6c14f998c53bb8ea009a4/SimPEG/potential_fields/gravity/simulation.py#L52

@lheagy 
Why did we switch to a for-loop here?

**Comment by jcapriot** [comment]
I think it was because you had wanted to avoid double storing the sensitivity matrix when it was in memory.

**Comment by lheagy** [comment]
Hmm, yeah I think @jcapriot's comment is right here. I think that the dense matrix multiplication grows memory quite a bit, but there might be a way to do that in-place? That isn't something I have looked into

**Comment by domfournier** [comment]
Oh yea totally, I recall now. The dask operation could do it in place, but now numpy.
Let's leave this guy open for a little while as a reminder.

**Comment by jcapriot** [comment]
Actually... I think the answer here to both questions, is `np.einsum`, and I believe the appropriate line of code will be,
```
np.einsum('i,ij,ij->j', W, G, G)
```

From some quick tests, this is even about 10 times faster than
```
np.sum(W[:, None] * (G * G), axis=0)
```

I still need to double check the memory requirements though.

**Comment by domfournier** [comment]
Wow that was quick. Thanks Eisten (@jcapriot  ;)

Just tried on this:

`import numpy as np
G = np.random.randn(10000,100000)
W = np.ones(G.shape[0])`
This indeed works:
`np.einsum('i,ij,ij->j', W, G, G)`

This blows up:
`np.sum(W[:, None] * G**2., axis=0)`

**Comment by jcapriot** [comment]
Yeah, `einsum` is actually very powerful and I love it, but I’m just still not the best at remembering _how_ to use it.

**Comment by jcapriot** [comment]
We still need to make more use of einsum for this operation in the code base.
